gangplank/ocp/ssh.go

```diff
@@ -37,107 +36,157 @@ func getSshMinioForwarder(j *spec.JobSpec) *SSHForwardPort {
+		callback := func() (string, error) {
+			pass, err := terminal.ReadPassword("Login password:")
+			return string(pass), err
 		}
+		authMethods = append(authMethods, ssh.PasswordCallback(callback))
+	}
 
-		cmd = exec.CommandContext(ctx, args[0], args[1:]...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		if err := cmd.Start(); err != nil {
-			return err
+	callback := ssh.InsecureIgnoreHostKey()
+	if secure {
+		if port != "22" {
+			host = fmt.Sprintf("[%s]:%s", host, port)
+		}
+		key := terminal.HostKey(host)
+		if key != nil {
+			callback = ssh.FixedHostKey(key)
 		}
-		// wait for SSH to start
-		time.Sleep(5 * time.Second)
-		return nil
 	}
 
-	// Give a first start to make sure it works
-	if err := run(); err != nil {
-		return nil, err
+	return ssh.Dial("tcp",
+		net.JoinHostPort(host, port),
+		&ssh.ClientConfig{
+			User:            user,
+			Auth:            authMethods,
+			HostKeyCallback: callback,
+			HostKeyAlgorithms: []string{
+				ssh.KeyAlgoRSA,
+				ssh.KeyAlgoDSA,
+				ssh.KeyAlgoECDSA256,
+				ssh.KeyAlgoECDSA384,
+				ssh.KeyAlgoECDSA521,
+				ssh.KeyAlgoED25519,
+			},
+			Timeout: 5 * time.Second,
+		},
+	)
+}
+
+// fowardOverSSH forwards the minio connection over SSH.
+func (m *minioServer) fowardOverSSH(termCh termChan, errCh chan<- error) error {
+	sshPort := 22
+	if m.overSSH.SSHPort != 0 {
+		sshPort = m.overSSH.SSHPort
 	}
+	sshport := strconv.Itoa(sshPort)
 
-	// Run and monitor ssh
-	go func() {
+	l := log.WithFields(log.Fields{
+		"remote host": m.overSSH.Host,
+		"remote user": m.overSSH.User,
+		"port":        m.Port,
+	})
 
-		// term kills the SSH forwarding command, and then unsets
-		term := func() {
-			log.Info("terminating ssh forwarder")
-			_ = syscall.Kill(cmd.Process.Pid, syscall.SIGTERM)
-			cmd = nil
-		}
+	l.Info("Forwarding local port over SSH to remote host")
 
-		// monitor ensures that the ssh command is running
-		monitor := func() <-chan bool {
-			done := make(chan bool)
-			go func() {
-				for {
-					select {
-					// terminate the monitor function if singalled
-					case t, ok := <-termCh:
-						if t || !ok {
-							return
-						}
-					// wait until the command terminates
-					default:
-						// do nothing if cmd has been undefined
-						if cmd == nil {
-							continue
-						}
-						// wait until cmd completes
-						_ = cmd.Wait()
-						done <- true
-					}
-				}
-			}()
-			return done
+	client, err := sshClient(m.overSSH.User, m.overSSH.Host, sshport, false, m.overSSH.Key)
+	if err != nil {
+		return err
+	}
+
+	// Open the remote port over SSH
+	remotePort, err := client.Listen("tcp4", fmt.Sprintf("127.0.0.1:%d", m.Port))
+	if err != nil {
+		err = fmt.Errorf("%w: failed to open remote port over sshfor proxy", err)
+		return err
+	}
+
+	// copyIO is a blind copier that copies between source and destination
+	copyIO := func(src, dest net.Conn) {
+		defer src.Close()  //nolint
+		defer dest.Close() //nolint
+		_, _ = io.Copy(src, dest)
```

Comment:
Shouldn't errors here forward to `errChan`?

Reply:
On a first pass, you would think so -- but propagating the `io.Copy` errors would be error prone since its at the application layer (did the client EOF, or did someone hit the local endpoint with curl?). Any errors on this will be reflected as IO errors on the client and will be propagated from the client and are more useful. Further, sending the error over the `errChan` will result in the whole job being scuttled before the other workers can return the work.